File: src/main/java/de/zalando/zally/rules/AvoidJavascriptKeywordsRule.java
package de.zalando.zally.rules;

import de.zalando.zally.Violation;
import de.zalando.zally.ViolationType;
import io.swagger.models.Model;
import io.swagger.models.Swagger;

import java.util.*;

public class AvoidJavascriptKeywordsRule implements Rule {

    public static final String TITLE = "Should: avoid reserved Javascript keywords";
    public static final String DESC_PATTERN = "Property name '%s' is reserved javascript keyword";
    public static final String RULE_URL = "http://zalando.github.io/restful-api-guidelines/naming/Naming.html" +
            "#should-reserved-javascript-keywords-should-be-avoided";
    private static final Set<String> RESERVED_KEYWORDS = new HashSet<>(Arrays.asList(
            "break", "do", "in", "typeof", "case", "else", "instanceof", "var", "catch", "export", "new", "void",
            "class", "extends", "return", "while", "const", "finally", "super", "with", "continue", "for", "switch",
            "yield", "debugger", "function", "this", "default", "if", "throw", "delete", "import", "try"
    ));

    @Override
    public List<Violation> validate(Swagger swagger) {
        List<Violation> res = new ArrayList<>();
        if (swagger.getDefinitions() != null) {
            for (Map.Entry<String, Model> entry : swagger.getDefinitions().entrySet()) {
                String definition = entry.getKey();
                if(entry.getValue().getProperties() != null) {
                    for (String prop : entry.getValue().getProperties().keySet()) {
                        if (RESERVED_KEYWORDS.contains(prop)) {
                            res.add(new Violation(
                                    TITLE,
                                    String.format(DESC_PATTERN, prop),
                                    ViolationType.SHOULD, RULE_URL, definition + "." + prop));
                        }
                    }
                }
            }
        }
        return res;
    }
}

File: server/src/main/java/de/zalando/zally/rules/AvoidJavascriptKeywordsRule.kt
package de.zalando.zally.rules

import de.zalando.zally.Violation
import de.zalando.zally.ViolationType
import de.zalando.zally.utils.getAllJsonObjects
import io.swagger.models.Swagger
import org.springframework.stereotype.Component

@Component
class AvoidJavascriptKeywordsRule : AbstractRule() {

    override val title = "Avoid reserved Javascript keywords"
    override val url = "http://zalando.github.io/restful-api-guidelines/naming/Naming.html" +
            "#should-reserved-javascript-keywords-should-be-avoided"
    override val violationType = ViolationType.SHOULD
    override val code = "S001"
    private val DESC_PATTERN = "Property names should not coincide with reserved javascript keywords"

    private val RESERVED_KEYWORDS = setOf(
            "break", "do", "in", "typeof", "case", "else", "instanceof", "var", "catch", "export", "new", "void",
            "class", "extends", "return", "while", "const", "finally", "super", "with", "continue", "for", "switch",
            "yield", "debugger", "function", "this", "default", "if", "throw", "delete", "import", "try"
    )

    override fun validate(swagger: Swagger): Violation? {
        val result = swagger.getAllJsonObjects().flatMap { (def, path) ->
            val badProps = def.keys.filter { it in RESERVED_KEYWORDS }
            if (badProps.isNotEmpty()) listOf(path + ": " + badProps.joinToString(", ") to path) else emptyList()
        }
        return if (result.isNotEmpty()) {
            val (props, paths) = result.unzip()
            Violation(this, title, DESC_PATTERN + "\n" + props.joinToString("\n"), violationType, url, paths)
        } else null
    }
}